presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/README.md

```diff
@@ -0,0 +1,148 @@
+# Presto planner and optimizer tests
+
+Testing planner and optimizer functionality generally requires generating a
+plan and then validating it against a set of expectations. The challenge is
+expressing the expectations in a way that's concise and resilient to changes in
+the objects Presto uses to represent the plan and the code that generates the
+plan.
+
+Unfortunately, that list includes the following:
+
+* The parser
+* The analyzer
+* The planner
+* Some or all of the optimizers
+* The plan objects themselves
+
+The plan testing framework attempts to solve the problem of specifiying
+expectations about query plans in tests.
+
+## Quickstart
+
+If you want to see examples of basic use of the plan testing framework, look at
+TestPlanFramework.java. If you'd like to try to begin writing new tests, look
+at PlanMatchPattern.java to see what sorts of patterns are already
+well-supported by the framework.
+
+### PlanMatchPatterns
+
+Expectations are expressed using `PlanMatchPatterns`. A `PlanMatchPattern`
+roughly corresponds to a `PlanNode`, in that it has zero or more
+`PlanMatchPatterns` as sources. A complete test generally expresses its
+expectations about an actual plan as a tree of `PlanMatchPatterns` that is
+structurally similar to the actual plan.
+
+A `PlanMatchPattern` by itself contains no expectations about the contents or
+even type of the node that it's meant to match. Instead, expectations are added
+to `PlanMatchPatterns` using `Matchers`.
+
+### Matchers
+
+`Matchers` are applied to `PlanNodes` to determine if the node they're applied
+to meets some expectation expressed in the `Matcher`.
+
+*A `Matcher` failing to match against a PlanNode it is applied to does not
+intrinsically mean that the plan does not match the expectations expressed by
+the test.* This is because the plan testing framework uses a pattern-matching
+like algorithm to allow tests to express their expectations flexibly with
+regards to changes in the structure of the tree by using the `anyTree()`
+pattern.  The `anyTree()` pattern looks for its child patterns in any of the
```

Comment:
pattern-matching-like
regard
looks for its child patterns in...the children...